### PR TITLE
push to stable

### PR DESCRIPTION
**Pull request by vanhauser-thc**

### CODE REVIEW

File: GNUmakefile
Comment (domenukk) [inline GNUmakefile]: Specific _to_ MacOS?

File: README.md
Comment (domenukk) [inline README.md:28]: `and up`?

File: docs/env_variables.md
Comment (domenukk) [inline docs/env_variables.md]: `-U` does not use `afl-qemu-trace`  (and I assume Frida mode doesn't, either?)

File: frida_mode/src/instrument.c
Comment (domenukk) [inline frida_mode/src/instrument.c:35]: Could this be improved by the neverzero asm we use elsewhere?
```c
#define INC_AFL_AREA(loc)           \
  asm volatile(                     \
      "addb $1, (%0, %1, 1)\n"      \
      "adcb $0, (%0, %1, 1)\n"      \
      : /* no out */                \
      : "r"(afl_area_ptr), "r"(loc) \
      : "memory", "eax")
```

File: frida_mode/src/instrument.c
Comment (domenukk) [inline frida_mode/src/instrument.c]: Why do stack ops matter for performance? I would assume stack access on par or faster than a global?

File: frida_mode/src/ranges.c
Comment (domenukk) [inline frida_mode/src/ranges.c]: That's an odd placement, probably from code formatting? Maybe empty brackets look better here?

File: frida_mode/test/testinstr.c
Comment (domenukk) [inline frida_mode/test/testinstr.c:80]: unchecked malloc

File: frida_mode/test/testinstr.c
Comment (domenukk) [inline frida_mode/test/testinstr.c:103]: what's the reason for this faux-loop?

File: frida_mode/test/testinstr.py
Comment (domenukk) [inline frida_mode/test/testinstr.py]: `#!/usr/bin/env python3` is (apparently) more portable.
Also I'd suggest running `black` on this file for formatting (should I do it?)

File: instrumentation/afl-llvm-lto-instrumentation.so.cc
Comment (domenukk) [inline instrumentation/afl-llvm-lto-instrumentation.so.cc:95]: Would be interested to know the story behind this change

File: qemu_mode/libcompcov/compcovtest.cc
Comment (domenukk) [inline qemu_mode/libcompcov/compcovtest.cc]: unchecked `fopen` return (and `argc` may not 2 in some future use?)

File: src/afl-fuzz-stats.c
Comment (domenukk) [inline src/afl-fuzz-stats.c]: Somehow feels odd. `custom-mutator-only mode` or just `custom mutators only`?

File: src/afl-fuzz-stats.c
Comment (domenukk) [inline src/afl-fuzz-stats.c:891]: Still default for `-M`. I would leave this decision to the compiler to optimize, maybe

**Review by domenukk — APPROVED** [review]

File: README.md
Comment (domenukk) [reply README.md:28]: not sure myself. I guess onward is okay :)

File: instrumentation/afl-llvm-lto-instrumentation.so.cc
Comment (vanhauser-thc) [reply instrumentation/afl-llvm-lto-instrumentation.so.cc:95]: otherwise there is a warning ... (types.h explains this)

File: src/afl-fuzz-stats.c
Comment (vanhauser-thc) [reply src/afl-fuzz-stats.c:891]: yes you would never run -M without at least one -S or usually more -S, so this is better.

File: frida_mode/src/instrument.c
Comment (vanhauser-thc) [reply frida_mode/src/instrument.c:35]: @WorksButNotTested

File: frida_mode/src/instrument.c
Comment (vanhauser-thc) [reply frida_mode/src/instrument.c]: @WorksButNotTested

File: frida_mode/test/testinstr.c
Comment (vanhauser-thc) [reply frida_mode/test/testinstr.c:103]: @WorksButNotTested

File: frida_mode/test/testinstr.c
Comment (vanhauser-thc) [reply frida_mode/test/testinstr.c:80]: @WorksButNotTested

File: frida_mode/src/ranges.c
Comment (vanhauser-thc) [reply frida_mode/src/ranges.c]: @WorksButNotTested

File: GNUmakefile
Comment (vanhauser-thc) [reply GNUmakefile]: "Consider using Linux or *BSD for fuzzing software not specific for MacOs." - that is correct

File: docs/env_variables.md
Comment (vanhauser-thc) [reply docs/env_variables.md]: right, will fix

File: src/afl-fuzz-stats.c
Comment (vanhauser-thc) [reply src/afl-fuzz-stats.c]: yeah "custom-mutator-only mode" is better

File: frida_mode/src/instrument.c
Comment (WorksButNotTested) [reply frida_mode/src/instrument.c:35]: I've now made this change in my own development branch for both x64 and Aarch64, but I've already spent a good amount of time optimizing the x64 assembly, so attempting to merge it in here is likely to be pretty messy. As this is a new feature, albeit small, rather than a defect are you happy we hold off on this until the next merge?

Comment (domenukk) [reply frida_mode/src/instrument.c:35]: Yes makes sense 👍

File: frida_mode/src/instrument.c
Comment (WorksButNotTested) [reply frida_mode/src/instrument.c]: I've clarified the comment. If tracing isn't used, then neither is the buffer, so we avoid a few precious instructions pushing/popping space on the stack.

File: frida_mode/src/ranges.c
Comment (WorksButNotTested) [reply frida_mode/src/ranges.c]: Fixed.

File: frida_mode/test/testinstr.c
Comment (WorksButNotTested) [reply frida_mode/test/testinstr.c:80]: Fixed

File: frida_mode/test/testinstr.c
Comment (WorksButNotTested) [reply frida_mode/test/testinstr.c:103]: I thought it was a fairly common pattern to allow clean-up in the event of failure if you don't have try/finally support. Happy to change it though if you have a preferred style for this sort of thing? Just point me at an example and I will copy it. Or maybe we don't need to care about cleaning up resources as we are about to exit the process anyway and it's only test code?

File: GNUmakefile
Comment (hexcoder-) [inline GNUmakefile:251]: should be SYS instead of ARCH

File: GNUmakefile
Comment (hexcoder-) [inline GNUmakefile:256]: dito

File: GNUmakefile
Comment (hexcoder-) [inline GNUmakefile:261]: dito

File: GNUmakefile
Comment (hexcoder-) [inline GNUmakefile:493]: dito

File: GNUmakefile
Comment (hexcoder-) [inline GNUmakefile:557]: could use also SYS instead of `uname`

File: src/afl-fuzz.c
Comment (hexcoder-) [inline src/afl-fuzz.c:237]: `SAYF` is a preprocessor macro. It is illegal to use preprocessor directives (`#ifdef`) inside macro arguments.
So this would have to be extracted out of here into its own preprocessor definition which can be used here (see other places, where I have extracted similar constructs)

File: utils/autodict_ql/build-codeql.sh
Comment (hexcoder-) [inline utils/autodict_ql/build-codeql.sh]: suggest a newline at the end here.

File: utils/autodict_ql/litool.ql
Comment (hexcoder-) [inline utils/autodict_ql/litool.ql]: Please add missing newline

File: utils/autodict_ql/memcmp-str.ql
Comment (hexcoder-) [inline utils/autodict_ql/memcmp-str.ql:8]: dito

File: utils/autodict_ql/readme.md
Comment (hexcoder-) [inline utils/autodict_ql/readme.md]: "this dict2file is automatically generates token extraction"

I do not understand that sentence. (I did a review on the text before)

**Review by hexcoder- — CHANGES_REQUESTED** [review]

File: utils/autodict_ql/readme.md
Comment (Microsvuln) [reply utils/autodict_ql/readme.md]: @hexcoder- Sorry, a typo, it actually should be "dict2file is automatically generates tokens" .

Comment (hexcoder-) [reply utils/autodict_ql/readme.md]: or maybe "dict2file is automatically generating tokens"?

File: utils/autodict_ql/memcmp-str.ql
Comment (Microsvuln) [reply utils/autodict_ql/memcmp-str.ql:8]: Add a newline in the end here also?

File: utils/autodict_ql/readme.md
Comment (Microsvuln) [reply utils/autodict_ql/readme.md]: Yeap, looks better :-)

File: utils/autodict_ql/memcmp-str.ql
Comment (hexcoder-) [reply utils/autodict_ql/memcmp-str.ql:8]: yes, please.

File: GNUmakefile
Comment (vanhauser-thc) [reply GNUmakefile:251]: oops thanks!

File: GNUmakefile
Comment (vanhauser-thc) [reply GNUmakefile:557]: true

File: src/afl-fuzz.c
Comment (vanhauser-thc) [reply src/afl-fuzz.c:237]: ah didnt know, fixing

**Review by hexcoder- — APPROVED** [review]

**Pull request merged**
